### PR TITLE
perf: create DocumentArray index map lazily

### PR DESCRIPTION
**Pull request by alaeddine-13**
closes https://github.com/jina-ai/internal-tasks/issues/279

results:
```python
from jina import Document
from jina.logging.profile import TimeContext

doc = Document()
with TimeContext('appending chunks'):
    for _ in range(10000):
        doc.chunks.append(
            Document(
                text='hey',
                offset=0,
                weight=1.0 ,
                location=[1, 2],
            )
        )
```

on master:
```
appending chunks ...	appending chunks takes 31 seconds (31.40s)
```

on feature branch:
```
appending chunks ...	appending chunks takes 0 seconds (0.64s)
```

### CODE REVIEW

File: jina/types/arrays/document.py [hunk 1]
```diff
@@ -83,7 +83,17 @@ def __init__(self, docs: Optional[DocumentArraySourceType] = None):
                     raise ValueError(
                         f'DocumentArray got an unexpected input {type(docs)}'
                     )
-        self._update_id_to_index_map()
+        self._id_to_index = None
+
+    @property
+    def id_to_index(self) -> Dict:
```
Comment (JoanFM) [inline jina/types/arrays/document.py:89]: make it provate

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/arrays/document.py (hunk 1 above)
Comment (alaeddine-13) [reply jina/types/arrays/document.py:89]: it's used in jina/types/document/graph.py so it can't be private (IDE used to show a warning)

Comment (hanxiao) [reply jina/types/arrays/document.py:89]: lets still make it private ignore graphDoc for now. graphDoc has been removed from our docs and will highly likely to be deprecated in 3.0

**Review by bwanglzu — APPROVED** [review]
LGTM! i also think we can safely ignore graph document at the moment

File: jina/types/arrays/document.py (hunk 1 above)
Comment (alaeddine-13) [reply jina/types/arrays/document.py]: changed

**Review by JoanFM — CHANGES_REQUESTED** [review]
Add a test in the types than make sure that a DocumentArray has `_id_to_index` to None at the beginning?

**Comment by alaeddine-13** [comment]
> Add a test in the types than make sure that a DocumentArray has `_id_to_index` to None at the beginning?

added

File: tests/unit/types/arrays/documentarray/test_documentarray.py [hunk 2]
```diff
@@ -344,3 +344,12 @@ def test_none_extend():
...
     da.extend(None)
     assert len(da) == 100
+
+
+def test_lazy_index_map():
+    da = DocumentArray([Document() for _ in range(100)])
+    assert da._id_to_index is None
+
+    # build index map
+    assert len(da._index_map.keys()) == 100
```
Comment (JoanFM) [inline tests/unit/types/arrays/documentarray/test_documentarray.py:354]: I would change the test to show that the `_index_map` is not `None` after accessing by `id`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/types/arrays/documentarray/test_documentarray.py (hunk 2 above)
Comment (alaeddine-13) [reply tests/unit/types/arrays/documentarray/test_documentarray.py]: done

**Review by JoanFM — APPROVED** [review]
LGTM! Great job!
